v2/yarpcrandpeer/list_test.go

```diff
@@ -0,0 +1,350 @@
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+			},
+		},
+		{
+			msg: "notify peer is now available",
+			retainedUnavailablePeerIDs: []string{"1"},
+			expectedAvailablePeers:     []string{"1"},
+			peerListActions: []yarpctest.PeerListAction{
+				yarpctest.UpdateAction{AddedPeerIDs: []string{"1"}},
+				yarpctest.ChooseAction{
+					InputContextTimeout: 10 * time.Millisecond,
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+				yarpctest.NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: yarpc.Available},
+				yarpctest.ChooseAction{ExpectedPeer: "1"},
+			},
+		},
+		{
+			msg: "notify peer is still available",
+			retainedAvailablePeerIDs: []string{"1"},
+			expectedAvailablePeers:   []string{"1"},
+			peerListActions: []yarpctest.PeerListAction{
+				yarpctest.UpdateAction{AddedPeerIDs: []string{"1"}},
+				yarpctest.ChooseAction{ExpectedPeer: "1"},
+				yarpctest.NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: yarpc.Available},
+				yarpctest.ChooseAction{ExpectedPeer: "1"},
+			},
+		},
+		// TODO: these next 2 have same msg as prior 2, but timeout initial
+		// choos, so not actually dupes, but unclear
+		{
+			msg: "notify peer is now unavailable", // TODO: these
+			retainedAvailablePeerIDs: []string{"1"},
+			expectedUnavailablePeers: []string{"1"},
+			peerListActions: []yarpctest.PeerListAction{
+				yarpctest.UpdateAction{AddedPeerIDs: []string{"1"}},
+				yarpctest.ChooseAction{ExpectedPeer: "1"},
+				yarpctest.NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: yarpc.Unavailable},
+				yarpctest.ChooseAction{
+					InputContextTimeout: 10 * time.Millisecond,
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+			},
+		},
+		{
+			msg: "notify peer is still unavailable",
+			retainedUnavailablePeerIDs: []string{"1"},
+			expectedUnavailablePeers:   []string{"1"},
+			peerListActions: []yarpctest.PeerListAction{
+				yarpctest.UpdateAction{AddedPeerIDs: []string{"1"}},
+				yarpctest.NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: yarpc.Unavailable},
+				yarpctest.ChooseAction{
+					InputContextTimeout: 10 * time.Millisecond,
+					ExpectedErr:         newUnavailableError(context.DeadlineExceeded),
+				},
+			},
+		},
+		{
+			msg: "notify invalid peer",
+			retainedAvailablePeerIDs: []string{"1"},
+			releasedPeerIDs:          []string{"1"},
+			peerListActions: []yarpctest.PeerListAction{
+				yarpctest.UpdateAction{AddedPeerIDs: []string{"1"}},
+				yarpctest.UpdateAction{RemovedPeerIDs: []string{"1"}},
+				yarpctest.NotifyStatusChangeAction{PeerID: "1", NewConnectionStatus: yarpc.Available},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			defer mockCtrl.Finish()
+
+			transport := yarpctest.NewMockDialer(mockCtrl)
+
+			// Healthy Transport Retain/Release
+			peerMap := yarpctest.ExpectPeerRetains(
```

Comment:
I wonder whether ExpectPeerRetain’s signature implies too much about the abstract peer list implementation that may not apply to all concrete peer list implementations.